freqtrade/freqai/base_models/PyTorchModelTrainer.py

```diff
@@ -0,0 +1,184 @@
+from pathlib import Path
+from typing import Any, Dict
+
+import pandas as pd
+import torch
+import torch.nn as nn
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader, TensorDataset
+
+
+logger = logging.getLogger(__name__)
+
+
+class PyTorchModelTrainer:
+    def __init__(
+            self,
+            model: nn.Module,
+            optimizer: Optimizer,
+            criterion: nn.Module,
+            device: str,
+            batch_size: int,
+            max_iters: int,
+            eval_iters: int,
+            init_model: Dict,
+            model_meta_data: Dict[str, Any] = {},
+    ):
+        """
+        :param model: The PyTorch model to be trained.
+        :param optimizer: The optimizer to use for training.
+        :param criterion: The loss function to use for training.
+        :param device: The device to use for training (e.g. 'cpu', 'cuda').
+        :param batch_size: The size of the batches to use during training.
+        :param max_iters: The number of training iterations to run.
+            iteration here refers to the number of times we call
+            self.optimizer.step(). used to calculate n_epochs.
+        :param eval_iters: The number of iterations used to estimate the loss.
+        :param init_model: A dictionary containing the initial model/optimizer
+         state_dict and model_meta_data saved by self.save() method.
+        :param model_meta_data: Additional metadata about the model (optional).
+        """
+        self.model = model
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.model_meta_data = model_meta_data
+        self.device = device
+        self.max_iters = max_iters
+        self.batch_size = batch_size
+        self.eval_iters = eval_iters
+
+        if init_model:
+            self.load_from_checkpoint(init_model)
+
+    def fit(self, data_dictionary: Dict[str, pd.DataFrame]):
+        """
+         - Calculates the predicted output for the batch using the PyTorch model.
+         - Calculates the loss between the predicted and actual output using a loss function.
+         - Computes the gradients of the loss with respect to the model's parameters using
+           backpropagation.
+         - Updates the model's parameters using an optimizer.
+        """
+        data_loaders_dictionary = self.create_data_loaders_dictionary(data_dictionary)
+        epochs = self.calc_n_epochs(
+            n_obs=len(data_dictionary["train_features"]),
+            batch_size=self.batch_size,
+            n_iters=self.max_iters
+        )
+        for epoch in range(epochs):
+            # evaluation
+            losses = self.estimate_loss(data_loaders_dictionary, data_dictionary)
+            logger.info(
+                f"epoch ({epoch}/{epochs}):"
+                f" train loss {losses['train']:.4f} ; test loss {losses['test']:.4f}"
+            )
+            # training
+            for batch_data in data_loaders_dictionary["train"]:
```

Comment:
I isnt `data_loaders_dictionary` filled with `dataloader` objects? Does this work to treat `dataloader` objects directly as batch data? 

It would be more classical to enumerate here on the train dataloader (which is precisely how you did it below in `estimate_loss()`).

Reply:
the train and eval are equivalent. 
the only difference is that in the eval i do it on both splits versos the train where i explictly only take train.

Reply:
Technically you enumerate on the dataloader object below, and you do not explicitly enumerate on the object here. 

As you mention, you do the same thing in both places, so for consistency and readability it is preferred to use the same syntax in both places to convey this point. Then I would lobby for explicitly enumerating on the dataloader here.

Reply:
i see what you mean, i'll add it.